### ISSUE TITLE
What is map(+1)?

### ISSUE THREAD

**Issue by hickford** [sharkdp/cube-composer#23]
I don't understand map(+1) 
1. Here it rotates the wall by one column to the left http://i.imgur.com/3hEXGQv.png 
2. Here it changes pink blocks to brown on the bottom row http://i.imgur.com/QzTkuHW.png

**Comment by sharkdp** [comment]
The levels in chapter 3 are different from the rest. The title of level 3.1 ("0b0 .. 0b111") was a subtle hint that the stacks in these levels are to be thought of as binary representations of the numbers from 0 (0 0 0, pink pink pink) to 7 (1 1 1, brown brown brown).

> Here it rotates the wall by one column to the left http://i.imgur.com/3hEXGQv.png

Furthermore, the mathematical operations (+1, ×2, ..) are evaluated modulo 8. This is why (+1) 'rotates' the numbers, 0 -> 1, 1 -> 2, ..., 7 -> 0 ([Cyclic group C_8](http://en.wikipedia.org/wiki/Cyclic_group), in case that makes anything more clear).

> Here it changes pink blocks to brown on the bottom row http://i.imgur.com/QzTkuHW.png

In this example, the numbers after the first step (×2) are 0 2 4 6 0 2 4 6. The subsequent (+1) flips the lowest bit: 1 3 5 7 1 3 5 7.

I realize that I'm asking quite a lot... mainly because it is still work in progress. I'm working on [adding help text to the different levels](https://github.com/sharkdp/cube-composer/compare/6fe0c57e8fd0ba1aee66b60cd5d08c322a4e779f...c60731fa5654b16068155c46add639efa8cfe410) right now, so hopefully this should be more clear in the future.

**Comment by hickford** [comment]
Thanks David, let me try again. Yes that makes sense, incrementing the same as rotating for the starting position.

Cool. Amusingly there's a class of games Towlr that deliberately don't tell how you play, that's part of the puzzle. eg. Terry Cavanagh's http://distractionware.com/games/flash/harmonilr/

**Comment by sharkdp** [comment]
I added a hint in the level 3.1 to make this a little bit easier.

I also spend more time than I would like to admit on this Towlr. I still haven't solved it but I will keep trying :smile:. Thanks for the link!

**Comment by RazvanCosmeanu** [comment]
This one was tricky to understand. I mapped the columns as values that corresponded with modulo 8 values of the resulted column binary: [8, 9, 10, 11, 12, 13, 14, 15](±8 here), then mapped the result array to get an ideea of how it should look like. After that, is was simpler than I thought initially. Loving this game. uhm.. perhaps not the best place to comment, but thanks for this!
